### PR TITLE
Add Decorators and tests

### PR DESCRIPTION
**Pull request by keithralphs**
- Add Decorators to enhance default SDK ones
- Streamline helper functions
- Include tests for new code
- Use updated copier template
- Target Python 3.12 as well

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -12,8 +12,8 @@ classifiers = [
 ]
 description = "A set of functions to reduce the boilerplate required to add observability to your service of module"
 dependencies = [
-    "fastapi[all]<0.99",
-    "typing_extensions<4.6",
+    "fastapi[all]",
+    "typing_extensions",
```
Comment (callumforrester) [inline pyproject.toml]: Now we're on Python3.10, `ParamSpec` is part of the standard library. You can remove this dependency and do

```python
from typing import ParamSpec
```

File: src/observability_utils/tracing/decorators.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+from typing import Any, Concatenate, TypeVar
+
+from opentelemetry.context import attach
+from opentelemetry.propagate import get_global_textmap
+from opentelemetry.trace import SpanKind, Tracer
+from typing_extensions import ParamSpec
+
+T = TypeVar("T")
+P = ParamSpec("P")
+SEPARATOR = "."
```
Comment (callumforrester) [inline src/observability_utils/tracing/decorators.py]: This constant should probably be private

```suggestion
_SEPARATOR = "."
```

File: src/observability_utils/tracing/decorators.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+def use_propagated_context(
+    func: Callable[P, T],
+) -> Callable[Concatenate[dict[str, Any] | None, P], T]:
+    """Retrieves the propagated context information from the carrier param which is
+    concatenated onto the target function and injects that into the local observablity
+    context.
+
+    Args:
+        func:The function to be decorated
```
Comment (callumforrester) [inline src/observability_utils/tracing/decorators.py]:
```suggestion
        func: The function to be decorated
```

**Review by callumforrester — CHANGES_REQUESTED** [review]
Only major change is the dependencies

File: src/observability_utils/tracing/decorators.py (hunk 2 above)
Comment (keithralphs) [reply src/observability_utils/tracing/decorators.py]: Done

File: src/observability_utils/tracing/decorators.py (hunk 3 above)
Comment (keithralphs) [reply src/observability_utils/tracing/decorators.py]: Done

**Review by callumforrester — APPROVED** [review]

**Pull request merged**